daml/ContingentClaims/Lifecycle.daml

```diff
@@ -78,20 +78,26 @@ lifecycle' _ _ (OneF id) = do
   pure $ One id
 lifecycle' _ _ (GiveF (_, m)) = local negate m
 lifecycle' spot _ (WhenF obs (c, m)) = do
-  wasLifecycled <- lift . lift $ eval spot obs
-  if wasLifecycled then m $> Zero else pure $ When obs c
+  predicate <- lift . lift $ eval spot obs
+  if predicate then m $> Zero else pure $ When obs c
```

Comment:
It's annoying that we still have to prune the `When` branch here, instead of in `pruneZeros'`.

I think the way to fix this (and in general to improve the behaviour greatly), is to refactor `lifecycle'` to be an unfold rather than a fold i.e. to traverse the tree top-down, and that way we'll also be able to skip irrelevant branches, getting a considerable performance boost.
We can then use `ghylo` to fuse the top-down `lifecycle'` with the bottom-up `pruneZeros` into a single O(n) pass. In theory, at least :wink:

Reply:
Also one curious thing is that as a result of this, the only thing that `pruneZeros'` operates on is the root nodes - i.e. those above `When` (if there are any at all), as anything below a `When` will be pruned by `lifecycle'` itself.